src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -381,6 +298,101 @@ internal static FileEncoding GetEncoding(string path)
+                object tmp = sessionState.PSVariable.GetValue("PSDefaultFileEncoding");
+                LanguagePrimitives.TryConvertTo<FileEncoding>(tmp, out encodingPreference);
+            }
+            catch
+            {
+                ;
+            }
+            return encodingPreference;
+        }
+
+        /// <summary>
+        /// Retrieve the encoding in a provider context
+        /// </summary>
+        public static Encoding GetProviderEncoding(CmdletProvider provider, FileEncoding encoding)
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = GetEncodingPreference(provider.SessionState);
+            // If the encoding isn't set, but is available as $PSDefaultFileEncoding, use that
+            // It the encoding is set use that, otherwise return the default encoding
+            if ( encoding == FileEncoding.Unknown && encodingPreference != FileEncoding.Unknown )
+            {
+                resolvedEncoding = GetEncoding(encodingPreference);
+            }
+            else if ( encoding != FileEncoding.Unknown )
+            {
+                resolvedEncoding = GetEncoding(encoding);
+            }
+            return resolvedEncoding;
+        }
+
+        // This is the way the encoding is implemented in PowerShell 5 and earlier.
+        // If the user sets the default encoding to WindowsLegacy, we will
+        // be able to encode for that
+        internal static Dictionary<String, Encoding> legacyEncodingMap =
+            new Dictionary<string, Encoding>(StringComparer.OrdinalIgnoreCase)
+            {
+                { "microsoft.powershell.commands.addcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportclixmlcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.exportcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportpssessioncommand", Encoding.UTF8 }, // with BOM
+                { "microsoft.powershell.commands.formathex", Encoding.ASCII },
+                { "microsoft.powershell.commands.newmodulemanifestcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.getcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.importcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.outfilecommand", Encoding.Unicode }, // This includes redirection
+                { "microsoft.powershell.commands.setcontentcommand", Encoding.ASCII },
+                // Providers are handled here
+                { "microsoft.powershell.commands.filesystemprovider", Encoding.ASCII },
+            };
+
+        /// Get the Windows legacy encoding from our encoding map
+        internal static Encoding GetWindowsLegacyEncoding(string name)
+        {
+            if ( legacyEncodingMap.ContainsKey(name))
```

Comment:
Can we use `return TryGetValue(name, out Encoding value) ? value : Encoding.Default`?